chronos/asyncsync.nim

```diff
@@ -285,19 +301,37 @@ proc popFirstNoWait*[T](aq: AsyncQueue[T]): T {.
   aq.popFirstImpl()
 
 proc popLastNoWait*[T](aq: AsyncQueue[T]): T {.
-    raises: [AsyncQueueEmptyError].} =
+     raises: [AsyncQueueEmptyError].} =
   ## Get an item from the end of the queue ``aq`` immediately.
   ##
   ## If queue ``aq`` is empty, then ``AsyncQueueEmptyError`` exception raised.
   if aq.empty():
     raise newException(AsyncQueueEmptyError, "AsyncQueue is empty!")
   aq.popLastImpl()
 
-proc addFirst*[T](aq: AsyncQueue[T], item: T) {.async: (raises: [CancelledError]).} =
+proc popFirstNoWaitSafe*[T](aq: AsyncQueue[T]): T {.
```

Comment:
looking at the usages, we could perhaps create a pop iterator instead? something like:

```iterator popItems(): T =
    yield item
    del item
```

Reply:
What about adding items until its not full?

Reply:
the goal this comment was to investigate if we can have an api that admit the risk of defects "naturally" - this is why I suggested a separate `UnboundedAsyncQueue` - its `add` would never raise.

I don't think that a program will want to switch between bounded or unbounded "dynamically" - I can understand that a program may want to control the size of a bounded queue, but it seems like we could separate the bounded and unbounded case using a separate type for unbounded - this unbounded queue would not raise exceptions on `add` - together with the `popItems` iterator (or some better name), we could have a "defect-free" API for unbounded queues that also doesn't have unwanted `raises` effects.

"safe" is a difficult term in general - raising panics is not a "Safe" thing to do, so it's hard to tell what is safe and what is not from looking at the word - it is not well defined in nim and therefore a source of confusion - better avoid defects altogether.